curvefs/src/client/s3/kvclient/kvclient.h

```diff
@@ -0,0 +1,199 @@
+        Stop();
+    }
+ private:
+    friend KvClient<MemCachedClient>;
+    bool InitImp() {
+        client_ = memcached(nullptr, 0);
+        return client_ != nullptr;
+    }
+
+    void StopImp() {
+        if (client_) {
+            memcached_free(client_);
+        }
+    }
+
+    bool InitImp(const std::string& config) {
+        client_ = memcached(config.c_str(), config.length());
+        return client_ != nullptr;
+    }
+
+    bool InitImp(const std::string& hostname, uint32_t port) {
+        client_ = memcached(nullptr, 0);
+        if (client_) {
+            res_ = memcached_server_add(client_, hostname.c_str(), port);
+            return res_ == MEMCACHED_SUCCESS;
+        }
+        return false;
+    }
+
+    bool SetImp(const std::string& key,
+                const char* value,
+                const int value_len,
+                std::string* errorlog) {
+        auto res = memcached_set(client_, key.c_str(), key.length(), value,
+                                 value_len, 0, 0);
+        if (res == MEMCACHED_SUCCESS) {
+            return true;
+        }
+        *errorlog = ResError(res);
+        return false;
+    }
+
+    bool GetImp(const std::string& key, std::string* value) {
+        uint32_t flags = 0;
+        size_t value_length = 0;
+        char* res = memcached_get(client_, key.c_str(), key.length(),
+                                  &value_length, &flags, &res_);
+        if (res != nullptr && value->empty()) {
+            value->reserve(value_length + 1);
+            value->assign(res, res + value_length + 1);
+            value->resize(value_length);
+            free(res);
+            return true;
+        }
+        return false;
+    }
+
+    bool AddServerImp(const std::string& hostname, const uint32_t port) {
+        server_ = memcached_server_list_append(server_, hostname.c_str(), port,
+                                               &res_);
+        return res_ == MEMCACHED_SUCCESS;
+    }
+
+    bool PushServerImp() {
+        res_ = memcached_server_push(client_, server_);
+        memcached_server_list_free(server_);
+        return res_ == MEMCACHED_SUCCESS;
+    }
+
+    int ServerCount() {
+        return static_cast<int>(memcached_server_count(client_));
+    }
+ public:
+    bool SetClientAttr(memcached_behavior_t flag, uint64_t data) {
+        return (memcached_success(memcached_behavior_set(client_, flag, data)));
+    }
+
+    uint64_t GetClientAttr(memcached_behavior_t flag) {
+        return memcached_behavior_get(client_, flag);
+    }
+
+    const std::string ResError() const {
+        return memcached_strerror(nullptr, res_);
+    }
+
+    // transform the res to a error string
+    const std::string ResError(const memcached_return_t res) {
+        return memcached_strerror(nullptr, res);
+    }
+
+    memcached_st* GetClient() { return client_; }
+
+ private:
+    memcached_return_t res_;
+    memcached_server_st* server_;
+    memcached_st* client_;
```

Comment:
> Because the memcached_st is not multithreaded safe, if different threads use it, memcached_ clone(NULL, client) is required. Here, we need to consider the problem of different thread calls.

fix.